R/write_output.R

```diff
@@ -0,0 +1,329 @@
+  # use NSE, so giving these a global binding. The standard dplyr hacks
+  # (str2lang, .data prefix) didn't work here because it's not dplyr and we're
+  # not accessing a dataframe.
+  imputed_reports <- obs_reports <- R <- r <- time <- NULL # nolint
+  stan_draws <- tidybayes::gather_draws(
+    stanfit,
+    imputed_reports[time],
+    obs_reports[time],
+    R[time],
+    r[time],
+  ) |>
+    data.table::as.data.table()
+
+  return(list(stan_draws = stan_draws, fact_table = fact_table))
+}
+
+#' Post-process and merge posterior draws with a fact table.
+#'
+#' This function merges posterior draws with a fact table containing
+#' date-time-parameter combinations. It also standardizes parameter names and
+#' renames key columns.
+#'
+#' @param draws A data.table of posterior draws (either raw or summarized).
+#' @param fact_table A data.table of unique date-time-parameter combinations.
+#'
+#' @return A data.table with merged posterior draws and standardized parameter
+#' names.
+#' @noRd
+post_process_and_merge <- function(
+    draws,
+    fact_table,
+    geo_value,
+    model,
+    disease) {
+  # Step 1: Left join the date-time-parameter map onto the Stan draws
+  merged_dt <- merge(
+    draws,
+    fact_table,
+    by = c("time", ".variable"),
+    all.x = TRUE,
+    all.y = FALSE
+  )
+
+  # Step 2: Standardize parameter names
+  data.table::set(merged_dt, j = ".variable", value = factor(
+    merged_dt[[".variable"]],
+    levels = c("imputed_reports", "obs_reports", "R", "r"),
+    labels = c("latent_cases", "obs_cases", "Rt", "growth_rate")
+  ))
+
+  # Step 3: Rename columns as necessary
+  data.table::setnames(
+    merged_dt,
+    old = c(
+      ".draw", ".chain", ".variable", ".value", ".lower", ".upper", ".width",
+      ".point", ".interval", "date", ".iteration"
+    ),
+    new = c(
+      "_draw", "_chain", "_variable", "value", "_lower", "_upper", "_width",
+      "_point", "_interval", "reference_date", "_iteration"
+    ),
+    # If using summaries, skip draws-specific names
+    skip_absent = TRUE
+  )
+
+  # Metadata for downstream querying without path parsing or joins
+  data.table::set(merged_dt, j = "geo_value", value = factor(geo_value))
+  data.table::set(merged_dt, j = "model", value = factor(model))
+  data.table::set(merged_dt, j = "disease", value = factor(disease))
+
+  return(merged_dt)
+}
+
+#' Process posterior samples from a Stan fit object (raw draws).
+#'
+#' Extracts raw posterior samples from a Stan fit object and post-processes
+#' them, including merging with a fact table and standardizing the parameter
+#' names. If calling `process_quantiles()` the 50% and 95% intervals are
+#' returned in `{tidybayes}` format.
+#'
+#' @param fit An EpiNow2 fit object with posterior estimates.
+#' @param disease,geo_value,model Metadata for downstream processing.
+#'
+#' @return A data.table of posterior draws or quantiles, merged and processed.
+#'
+#' @name sample_processing_functions
+NULL
+
```

Comment:
What's this do? Not familiar with it

Reply:
https://cran.r-project.org/web/packages/roxygen2/vignettes/reuse.html